stdlib/LinearAlgebra/src/adjtrans.jl

```diff
@@ -378,15 +378,31 @@ Broadcast.broadcast_preserving_zero_d(f, tvs::Union{Number,TransposeAbsVec}...)
 
 
 ### reductions
-# faster to sum the Array than to work through the wrapper
-Base._mapreduce_dim(f, op, init::Base._InitialValue, A::Transpose, dims::Colon) =
+# faster to sum the Array than to work through the wrapper (but only in commutative reduction ops)
+const CommutativeOps = Union{typeof(+),typeof(Base.add_sum),typeof(-),typeof(min),typeof(max),typeof(|),typeof(&)}
```

Comment:
I haven't looked any closer at the code where this is used, but having `-` in a list with this name sounds wrong.

```suggestion
const CommutativeOps = Union{typeof(+),typeof(Base.add_sum),typeof(min),typeof(max),typeof(|),typeof(&)}
```
But as `-` is also not associative, it should not be used with `reduce` anyway I guess.

Reply:
The name is bad, I agree, but in the case of reduction, we have

```julia
init - f(a_1) - f(a_2) - ... - f(a_n) = init - (f(a_1) + f(a_2) + ... + f(a_n))
```

where the terms in the last bracket commute. So, by "`CommutativeOps` we don't mean `op(x, y) = op(y, x)`, but rather `op(op(x, y), z) = op(op(x, z), y)`. I'll add a comment to the very least, and will keep thinking about a better name for the union. Perhaps `CommutativeReductionOps`... 😅

Reply:
From what I understand, yes:

https://github.com/JuliaLang/julia/blob/81eb6ef068fe2999340b4403a3855f38e361b74f/base/reducedim.jl#L362-L363

Reply:
But that's an implementation detail, not something that is guaranteed, or?

Reply:
Which example do you have in mind that would be critical in connection with `-`? I'm not insisting on including `-` in the list, I'd just like to understand which possible issues can arise... out of curiosity. To me, `-` is like `+`, which we include in the list, with a prepended `-`, so I don't see the issue, but may be missing something.

Reply:
```julia
julia> foldl(-, 1:4)
-8

julia> foldr(-, 1:4)
-2

julia> reduce(-, 1:4)
-8 # could also be -2 or (1-2)-(3-4)=0 or 1-((2-3)-4)=6
```
Adding commutations in there would probably not make things much worse, but feels strange to explicitly allow it.

Reply:
I see. I think that in Base, both the generator-based and the array specific implementation operate in the `foldl` way, currently. But I see the point that it is not specified/guaranteed. Thanks for taking the time to explain!